Implement exact source (xs) for magnet URIs

Sebmaster opened the pull request:
Since I saw #714 stalled a bit, I wanted to check in if no one's working on this and started off with a minimal working prototype.

I know I still need to move the fixture over (also don't rely on webtorrent.io, but create a local server instead) to the other repo, but does this look okay / is anyone else working on this?

feross commented on lib/torrent.js:
```diff
@@ -269,6 +271,34 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
+    try {
+      this._xsRequest = get.concat(opts, function (err, res, torrent) {
+        if (err || !torrent.length) return
+
```
Should check that `res.statusCode === 200`

feross commented:
Just took a quick look. Looks pretty good to me. I don't think we to/can easily support ftp:// so I would remove that test case.

Sebmaster commented:
Fixed. 

> I don't think we to/can easily support ftp:// so I would remove that test case.

I added that mainly to check if unsupported protocols would crash the process, since the spec says to silently ignore.

feross commented:
I think we should check for http:// https:// explicitly before passing into `simple-get`. That would be more reliable and we could get rid of the try-catch around `simple-get`, right?

Sebmaster commented:
Done.

> That would be more reliable and we could get rid of the try-catch around `simple-get`, right?

For the protocol case yeah, just hope there's nothing else `http` will throw on...

Sebmaster commented:
Okay, also migrated the tests over to a local server and reused the fixtures from `webtorrent-fixtures`. Should be ready (for xs, at least).

feross commented:
Thanks -- will take a look after getting WebTorrent Desktop v0.5 out.

Sebmaster commented:
I just checked the malformed URL case some more - a URL like `http://example.org:2455665555/` would now crash, which isn't great, so I'm thinking leaving the try / catch there would be better (either instead, or in addition to the protocol check).

feross commented on lib/torrent.js:
```diff
@@ -269,6 +271,38 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
+
+  if (parsedTorrent.xs && !self.info) {
+    // TODO: Could try multiple in parallel here
+    var src = parsedTorrent.xs instanceof Array ? parsedTorrent.xs[0] : parsedTorrent.xs
```
`Array.isArray()` is the right way to check if something is an Array.

feross commented on lib/torrent.js:
```diff
@@ -269,6 +271,38 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
+        }
+      }
+      this._xsRequest = get.concat(opts, function (err, res, torrent) {
+        if (err || res.statusCode !== 200 || !torrent.length) return
```
No need to check for `!torrent.length` since that will make `parseTorrent` throw below anyway.

feross replied:
Would be nice to add a log here with `debug` for debugging purposes.

feross commented on lib/torrent.js:
```diff
@@ -269,6 +271,38 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
+
+        if (parsedTorrent) {
+          if (parsedTorrent.infoHash !== self.infoHash) {
+            return
```
Debug log here too

feross commented on lib/torrent.js:
```diff
@@ -371,6 +405,10 @@ Torrent.prototype._onListening = function () {
+  if (self._xsRequest) {
+    self._xsRequest.abort()
+    self._xsRequest = null
+  }
```
This should go after the next line.

feross commented on test/node/download-metadata.js:
```diff
@@ -0,0 +1,65 @@
+
+  createServer(fixtures.leaves.torrent, function (url, next) {
+    client.add(fixtures.leaves.magnetURI + '&xs=' + encodeURIComponent(url), function (torrent) {
+      t.equal(torrent.name, 'Leaves of Grass by Walt Whitman.epub')
```
The name is also in the magnet URI as the `&dn` parameter. Might be better to check for another parameter.

feross commented on test/node/download-metadata.js:
```diff
@@ -0,0 +1,65 @@
+  var client = new WebTorrent({ dht: false, tracker: false })
+  client.add(fixtures.leaves.magnetURI + '&xs=' + encodeURIComponent('invalidurl:example'))
+  setTimeout(function () {
+    t.ok(true, 'no crash')
```
No need for this. The tests will fail if there's a thrown exception. We should add listeners for 'error' and 'warning' events to all these tests though.

feross replied:
Nevermind, I see what you were doing.

Sebmaster commented:
I totally fixed all of those just now :smile: 

Also what about the try-catch situation with malformed urls?

feross commented:
Ah, thanks. I'm fixing all that up too. Sorry, should have communicated that better.

feross commented:
I'm adding support for multiple &xs= params too.

feross commented:
@Sebmaster See https://github.com/feross/webtorrent/pull/809
